source/common/http/http1/parser.h

```diff
@@ -0,0 +1,183 @@
+
+  /**
+   * Called when header field data is received.
+   * @param data supplies the start address.
+   * @param length supplies the length.
+   * @return Status representing success or failure.
+   */
+  virtual Status onHeaderField(const char* data, size_t length) PURE;
+
+  /**
+   * Called when header value data is received.
+   * @param data supplies the start address.
+   * @param length supplies the length.
+   * @return Status representing success or failure.
+   */
+  virtual Status onHeaderValue(const char* data, size_t length) PURE;
+
+  /**
+   * Called when headers are complete. A base routine happens first then a virtual dispatch is
+   * invoked. Note that this only applies to headers and NOT trailers. End of
+   * trailers are signaled via onMessageCompleteBase().
+   * @return An error status or a ParserStatus.
+   */
+  virtual Envoy::StatusOr<ParserStatus> onHeadersComplete() PURE;
+
+  /**
+   * Called when body data is received.
+   * @param data supplies the start address.
+   * @param length supplies the length
+   */
+  virtual void bufferBody(const char* data, size_t length) PURE;
+
+  /**
+   * Called when the HTTP message has completed parsing.
+   * @return An error status or a ParserStatus.
+   */
+  virtual StatusOr<ParserStatus> onMessageComplete() PURE;
+
+  /**
+   * Called when accepting a chunk header.
+   */
+  virtual void onChunkHeader(bool) PURE;
+
+  virtual int setAndCheckCallbackStatus(Status&& status) PURE;
+  virtual int setAndCheckCallbackStatusOr(Envoy::StatusOr<ParserStatus>&& statusor) PURE;
+};
+
+class Parser {
+public:
+  // Struct containing the return value from parser execution.
+  struct RcVal {
+    // Number of parsed bytes.
+    size_t nread;
+    // Integer error from parser indicating return code.
+    int rc;
+  };
+  virtual ~Parser() = default;
+
+  // Executes the parser.
+  // @return an RcVal containing the number of parsed bytes and return code.
+  virtual RcVal execute(const char* slice, int len) PURE;
+
+  // Unpauses the parser.
+  virtual void resume() PURE;
+
+  // Pauses the parser and returns a status indicating pause.
+  virtual ParserStatus pause() PURE;
+
+  // Returns a parser status representing the errno value from the parser.
+  virtual ParserStatus getStatus() PURE;
+
+  // Returns an integer representing the status code stored in the parser structure. For responses
+  // only.
+  virtual uint16_t statusCode() const PURE;
+
+  // Returns an integer representing the HTTP major version.
+  virtual int httpMajor() const PURE;
+
+  // Returns an integer representing the HTTP minor version.
+  virtual int httpMinor() const PURE;
+
+  // Returns the number of bytes in the body. -1 if no Content-Length header
+  virtual uint64_t contentLength() const PURE;
+
+  // Returns parser flags (e.g. chunked).
+  virtual int flags() const PURE;
+
+  // Returns an integer representing the method. For requests only.
+  virtual uint16_t method() const PURE;
```

Comment:
are these going to be consistent across parsers?  Better t ohave
isPostMethod() or isMethod(Http::Headers::get().MethodValues.Post)?

Reply:
Removed this in favor of returning a string and comparing with `MethodValues` per the other comment about the enum class. 

Still open to introduce `isMethod(_)`, which I like (we still need the method name in the codec)